File: chapter1/006/main.go
package main

import (
	"fmt"
	"strings"
)

func main() {
	fmt.Println(compression("aabcccccaaa"))

	fmt.Println(compression("aaa"))
	fmt.Println(compression("aa"))
	fmt.Println(compression("aabb"))
	fmt.Println(compression(""))
	fmt.Println(compression("abcd"))
	fmt.Println(compression("abcdddddd"))
	fmt.Println(compression("aaAA"))
	fmt.Println(compression("aaaAAA"))
}

func compression(s string) string {
	newStringSlice := make([]string, len(s))

	lastRune := '-'
	var count int
	for _, r := range s {
		if lastRune == r || lastRune == '-' {
			count++
		} else {
			newStringSlice = append(newStringSlice, fmt.Sprintf("%s%d", string(lastRune), count))
			count = 1
		}
		lastRune = r
	}
	if count > 0 {
		newStringSlice = append(newStringSlice, fmt.Sprintf("%s%d", string(lastRune), count))
	}

	newString := strings.Join(newStringSlice, "")
	if len(newString) < len(s) {
		return newString
	}

	return s
}

File: chapter1/006/main.py
def main():
    print(compression("aabcccccaaa"))

    print(compression("aaa"))
    print(compression("aa"))
    print(compression("aabb"))
    print(compression(""))
    print(compression("abcd"))
    print(compression("abcdddddd"))
    print(compression("aaAA"))
    print(compression("aaaAAA"))

def compression(string):
    if len(string) == 0:
        return ""

    buffer = []

    count = 0
    lastChar = string[0]
    for c in string:
        if lastChar != c:
            buffer.append(f'{lastChar}{count}')
            count = 1
        else:
            count += 1

        lastChar = c

    if count > 0:
        buffer.append(f'{lastChar}{count}')
    
    new_string = "".join(buffer)

    if len(string) <= len(new_string):
        return string
    else:
        return new_string

main()